core/src/sigverify_shreds.rs

```diff
@@ -0,0 +1,395 @@
+fn verify_shred_cpu(packet: &Packet, slot_leaders: &HashMap<u64, Pubkey>) -> Option<u8> {
+    let sig_start = 0;
+    let sig_end = size_of::<Signature>();
+    let slot_start = sig_end + size_of::<ShredType>();
+    let slot_end = slot_start + size_of::<u64>();
+    let msg_start = sig_end;
+    let msg_end = packet.meta.size;
+    trace!("slot start and end {} {}", slot_start, slot_end);
+    let slot: u64 = deserialize(&packet.data[slot_start..slot_end]).ok()?;
+    trace!("slot {}", slot);
+    let pubkey: &Pubkey = slot_leaders.get(&slot)?;
+    let signature = Signature::new(&packet.data[sig_start..sig_end]);
+    trace!("signature {}", signature);
+    if !signature.verify(pubkey.as_ref(), &packet.data[msg_start..msg_end]) {
+        return Some(0);
+    }
+    Some(1)
+}
+
+pub fn verify_shreds_cpu(batches: &[Packets], slot_leaders: &HashMap<u64, Pubkey>) -> Vec<Vec<u8>> {
+    use rayon::prelude::*;
+    let count = batch_size(batches);
+    debug!("CPU SHRED ECDSA for {}", count);
+    let rv = PAR_THREAD_POOL.with(|thread_pool| {
+        thread_pool.borrow().install(|| {
+            batches
+                .into_par_iter()
+                .map(|p| {
+                    p.packets
+                        .par_iter()
+                        .map(|p| verify_shred_cpu(p, slot_leaders).unwrap_or(0))
+                        .collect()
+                })
+                .collect()
+        })
+    });
+    inc_new_counter_debug!("ed25519_shred_verify_cpu", count);
+    rv
+}
+
+fn shred_gpu_pubkeys(
+    batches: &[Packets],
+    slot_leaders: &HashMap<u64, Pubkey>,
+    recycler_offsets: &Recycler<TxOffset>,
+    recycler_pubkeys: &Recycler<PinnedVec<Pubkey>>,
+) -> (PinnedVec<Pubkey>, TxOffset) {
+    //TODO: mark Pubkey::default shreds as failed after the GPU returns
+    assert_eq!(slot_leaders.get(&std::u64::MAX), Some(&Pubkey::default()));
+    let slots: Vec<Vec<u64>> = PAR_THREAD_POOL.with(|thread_pool| {
+        thread_pool.borrow().install(|| {
+            batches
+                .into_par_iter()
+                .map(|p| {
+                    p.packets
+                        .iter()
+                        .map(|packet| {
+                            let slot_start = size_of::<Signature>() + size_of::<ShredType>();
+                            let slot_end = slot_start + size_of::<u64>();
+                            let slot: Option<u64> =
+                                deserialize(&packet.data[slot_start..slot_end]).ok();
+                            if slot.is_some() && slot_leaders.get(&slot.unwrap()).is_some() {
+                                slot.unwrap()
+                            } else {
+                                std::u64::MAX
+                            }
+                        })
+                        .collect()
+                })
+                .collect()
+        })
+    });
+    let mut keys: HashMap<Pubkey, Vec<u64>> = HashMap::new();
```

Comment:
nits: I prefer a more describe name like `keys_to_slots`, following the style of below `slot_to_key_ix`.